build failure: hpx::string_util::trim()

```
$ cmake3 -DCMAKE_INSTALL_PREFIX=/opt/hpx -DHPX_WITH_MALLOC=system -DHPX_WITH_FETCH_ASIO=ON -DHWLOC_ROOT=/opt/hwlock ..
$ gcc --version
gcc (GCC) 10.2.1 20200804 (Red Hat 10.2.1-2)
$ make
...
In file included from /home/user/hpx/build/libs/core/string_util/include/hpx/modules/string_util.hpp:17,
                 from /home/user/hpx/libs/full/include/include/hpx/include/util.hpp:43,
                 from /home/user/hpx/examples/accumulators/accumulator_client.cpp:15:
/home/user/hpx/libs/core/string_util/include/hpx/string_util/trim.hpp: In instantiation of ‘void hpx::string_util::trim(std::basic_string<_CharT, _Traits, _Alloc>&) [with CharT = char; Traits = std::char_traits<char>; Alloc = std::allocator<char>]’:
/home/user/hpx/examples/accumulators/accumulator_client.cpp:48:40:   required from here
/home/user/hpx/libs/core/string_util/include/hpx/string_util/trim.hpp:19:16: error: no matching function for call to ‘std::basic_string<char>::erase(std::basic_string<char>::iterator, __gnu_cxx::__normal_iterator<const char*, std::basic_string<char> >&)’
   19 |         s.erase(std::begin(s), first);
      |         ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
In file included from /opt/rh/devtoolset-10/root/usr/include/c++/10/string:55,
                 from /home/user/hpx/libs/core/errors/include/hpx/errors/error.hpp:15,
                 from /home/user/hpx/build/libs/core/errors/include/hpx/modules/errors.hpp:14,
                 from /home/user/hpx/libs/full/init_runtime/include/hpx/hpx_finalize.hpp:12,
                 from /home/user/hpx/libs/full/init_runtime/include/hpx/hpx_init.hpp:15,
                 from /home/user/hpx/examples/accumulators/accumulator_client.cpp:13:
/opt/rh/devtoolset-10/root/usr/include/c++/10/bits/basic_string.h:4700:7: note: candidate: ‘std::basic_string<_CharT, _Traits, _Alloc>& std::basic_string<_CharT, _Traits, _Alloc>::erase(std::basic_string<_CharT, _Traits, _Alloc>::size_type, std::basic_string<_CharT, _Traits, _Alloc>::size_type) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>; std::basic_string<_CharT, _Traits, _Alloc>::size_type = long unsigned int]’
 4700 |       erase(size_type __pos = 0, size_type __n = npos)
      |       ^~~~~
/opt/rh/devtoolset-10/root/usr/include/c++/10/bits/basic_string.h:4700:23: note:   no known conversion for argument 1 from ‘std::basic_string<char>::iterator’ to ‘std::basic_string<char>::size_type’ {aka ‘long unsigned int’}
 4700 |       erase(size_type __pos = 0, size_type __n = npos)
      |             ~~~~~~~~~~^~~~~~~~~
/opt/rh/devtoolset-10/root/usr/include/c++/10/bits/basic_string.h:4716:7: note: candidate: ‘std::basic_string<_CharT, _Traits, _Alloc>::iterator std::basic_string<_CharT, _Traits, _Alloc>::erase(std::basic_string<_CharT, _Traits, _Alloc>::iterator) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>; std::basic_string<_CharT, _Traits, _Alloc>::iterator = std::basic_string<char>::iterator]’
 4716 |       erase(iterator __position)
      |       ^~~~~
/opt/rh/devtoolset-10/root/usr/include/c++/10/bits/basic_string.h:4716:7: note:   candidate expects 1 argument, 2 provided
/opt/rh/devtoolset-10/root/usr/include/c++/10/bits/basic_string.h:4736:7: note: candidate: ‘std::basic_string<_CharT, _Traits, _Alloc>::iterator std::basic_string<_CharT, _Traits, _Alloc>::erase(std::basic_string<_CharT, _Traits, _Alloc>::iterator, std::basic_string<_CharT, _Traits, _Alloc>::iterator) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>; std::basic_string<_CharT, _Traits, _Alloc>::iterator = std::basic_string<char>::iterator]’
 4736 |       erase(iterator __first, iterator __last);
      |       ^~~~~
/opt/rh/devtoolset-10/root/usr/include/c++/10/bits/basic_string.h:4736:40: note:   no known conversion for argument 2 from ‘const char* const’ to ‘std::basic_string<char>::iterator’
 4736 |       erase(iterator __first, iterator __last);
      |                               ~~~~~~~~~^~~~~~
```

hpx 1.7.1

--- a/libs/core/string_util/include/hpx/string_util/trim.hpp
+++ b/libs/core/string_util/include/hpx/string_util/trim.hpp
@@ -14,11 +14,21 @@ namespace hpx { namespace string_util {
     template <typename CharT, class Traits, class Alloc>
     void trim(std::basic_string<CharT, Traits, Alloc>& s)
     {
+        // When using the pre-C++11 ABI in libstdc++, basic_string::erase
+        // does not have an overload taking different begin and end iterators.
+#if defined(_GLIBCXX_USE_CXX11_ABI) && _GLIBCXX_USE_CXX11_ABI == 0
+        auto first = std::find_if_not(std::begin(s), std::end(s),
+#else
         auto first = std::find_if_not(std::cbegin(s), std::cend(s),
+#endif
             [](int c) { return std::isspace(c); });
         s.erase(std::begin(s), first);
 
+#if defined(_GLIBCXX_USE_CXX11_ABI) && _GLIBCXX_USE_CXX11_ABI == 0
+        auto last = std::find_if_not(std::rbegin(s), std::rend(s),
+#else
         auto last = std::find_if_not(std::crbegin(s), std::crend(s),
+#endif
             [](int c) { return std::isspace(c); });
         s.erase(last.base(), std::end(s));
     }

